feat: LEAP-1973: Adjustable spans for Text tag

hlomzik opened the pull request:
Allow users to resize selected NER regions.


https://github.com/user-attachments/assets/1e6db970-6719-4edd-ad30-fb0b5fdc020a



The main idea is to use browser selection with styles of the region so the region expansion will look
smooth and natural. At the beginning the selection will be set to wrap the region. So users will just change
the selection during drag-n-drop, no extra code required for the visual part.

Steps to achieve this:
1. Detect that user is about to resize the region (mousedown on a handle)
2. Set selection style to the region's style
3. Set selection range to the region's range for the initial state
4. We do nothing on mousemove, and on mouseup we update the region's offsets
5. Remove selection style

There are some tricky parts here: selection should be created after we started dragging,
if we create it before, then the drag will drag-n-drop the selection instead of continuing it.
Steps and flags on init:
- mousedown: check if we are on a handle, set `draggableRegion` and `dragBackwards` to understand the direction
- mousemove: if we have `draggableRegion` but don't have `initializedDrag`, set it; that means we are dragging
- mousemove: if we have both, create selection around the region, set `currentSelection`;
             we are editing the region now
And we should reset all of them on mouseup.

hlomzik committed:
feat: LEAP-1973: Adjustable spans for Text tag

hlomzik committed:
Merge branch 'develop' into fb-leap-1973/adjustable-spans

hlomzik committed:
Fix visual glitches and issues in multiline text

- no weird selection stays on the screen
- no leftover spans after adjustment in multiline
- correct offsets in all cases
- better naming
- no extra regions created by accident

hlomzik committed:
Fix spans resetting during removing highlights

hlomzik committed:
Temporary style fix

- remove overflow to show handles in a first row
- adjust line-height to remove the gap between lines

hlomzik committed:
Move styles for resize handles to a scss to be general ones

This will improve performance and DX as they will be always in one place.

hlomzik committed:
Create only one undo/redo item on region resize

hlomzik committed:
Update region highlight on undo/redo

Internals updated with no problems, so we had to solve two issues with visuals:
- detect changes; for this we store offsets in spans and compare them during visuals update
- update hightlight by recreating spans

`_spans` are added to volatile props to trigger MST views updates.

hlomzik committed:
Update stored text on every region resize

`force` option was added to `updateHighlightedText()` to update text every time

hlomzik committed:
Handle undo/redo issues: just freeze it

hlomzik committed:
Fix linting

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14311504047)
> Successfully merged:  create mode 100644 web/libs/ui/src/shadcn.ts

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1973/adjustable-spans'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14311504047

bmartel commented on web/libs/editor/src/tags/object/RichText/RichText.scss:
```diff
@@ -2,9 +2,9 @@
 
   &__container {
-    overflow: auto;
+    // overflow: auto;
```
Should be removed if not needed

hlomzik replied:
rewrote it with FF just in case, because it might change the overall experience with Text tag. thank you!

bmartel commented on web/libs/editor/src/tags/object/RichText/RichText.scss:
```diff
@@ -83,8 +83,59 @@
+        position: absolute;
+        width: $handle_r * 2 + 1px;
+        height: $handle_r * 2 + 1px;
+        background: var(--grape_500);
```
Should be using a semantic CSS var

bmartel replied:
```suggestion
        background: var(--color-primary-content-subtle);
```

bmartel commented on web/libs/editor/src/tags/object/RichText/RichText.scss:
```diff
@@ -83,8 +83,59 @@
+        content: '';
+        width: 1px;
+        height: 100%;
+        background: var(--grape_500);
```
Should be using a semantic CSS var

bmartel replied:
```suggestion
        background: var(--color-primary-content-subtle);
```

bmartel commented on web/libs/editor/src/tags/object/RichText/RichText.scss:
```diff
@@ -83,8 +83,59 @@
+        bottom: -($handle_r * 2 + 1px);
+      }
+      &:hover::after {
+        outline: 1px solid var(--grape_500);
```
Should be using a semantic CSS var

bmartel replied:
```suggestion
        outline: 1px solid var(--color-primary-border-subtle);
```

bmartel commented:
Could you add a screenshot of what this looks like?

Gondragos commented on web/libs/editor/src/mixins/HighlightMixin.js:
```diff
@@ -73,19 +80,32 @@ export const HighlightMixin = types
+        // update label tag (index + attached labels) which sits in the last span
         Utils.Selection.applySpanStyles(lastSpan, { index: self.region_index, label: self.getLabels() });
+        // store offsets in spans for further comparison if region got resized
+        firstSpan.setAttribute("data-start", offsets.start);
```
Most likely this can be stored in a volatile variable or defined inside `.views(() => ({}))`, to avoid polluting the DOM.

Gondragos commented on web/libs/editor/src/mixins/HighlightMixin.js:
```diff
@@ -101,20 +121,49 @@ export const HighlightMixin = types
+      classes.forEach((resizeClass, index) => {
+        // html element that can't be encountered in a usual html
+        const handleArea = document.createElement("area");
+
```
It seems more efficient to store `handleArea` in the model's internal data, so that when using `querySelectorAll`, no more elements than planned are removed.

Gondragos commented on web/libs/editor/src/tags/object/RichText/RichText.scss:
```diff
@@ -83,8 +83,59 @@
+  :global(.htx-highlight) {
+    // resize handles around selected region,
+    // they look like popsicles, normal one at the beginning, upside down one at the end
+    area {
```
It might be better to rely on `__resize_right` and `__resize_left` instead of the tag name. Just in case the tag appears in the content being marked up.

Gondragos commented on web/libs/editor/src/tags/object/RichText/view.jsx:
```diff
@@ -51,12 +52,205 @@ class RichTextPieceView extends Component {
+   * @param {Document} doc document to apply style to
+   */
+  _setSelectionStyle = (region, doc) => {
+    const styleMap = region._spans[0].computedStyleMap();
```
Most likely, we have the necessary data within the model itself.

Gondragos commented on web/libs/editor/src/tags/object/RichText/view.jsx:
```diff
@@ -51,12 +52,205 @@ class RichTextPieceView extends Component {
+    const target = ev.target;
+    const region = this._determineRegion(target);
+    const classes = [STATE_CLASS_MODS.leftHandle, STATE_CLASS_MODS.rightHandle];
+    const isHandle = target.classList.contains(classes[0]) || target.classList.contains(classes[1]);
```
If you could store references to handle, you can compare the refs instead of the classes.

hlomzik committed:
Adjust colors of region and selection for better visuals

Now it's easy to see extended/reduced parts of the region.
The trick is to use lighter color for both region and selection,
and overlapped area will look darker, the same as the original region color.

hlomzik committed:
Handle dragging outside of the object tag

If it's outside just do nothing to the region, but stop the resizing and reset flags and styles.
To achieve this we have to attach handler to the document and do some tweaks.

hlomzik committed:
Add FF for the feature

hlomzik committed:
Better FF usage

Add `canResizeSpans` getter to RichText model and use it

hlomzik committed:
Update tag styles only with FF on

hlomzik committed:
Use semantic css vars

Co-authored-by: bmartel <brandonmartel@gmail.com>

Gondragos commented on web/libs/editor/src/mixins/HighlightMixin.js:
```diff
@@ -29,30 +33,44 @@ export const HighlightMixin = types
       `;
     },
+    get styles() {
+      return this.generateStyles(self.className, self.getColors());
```
It looks like styles and resizeStyles are the same thing and could simply be recalculated based on an internal volatile flag instead of creating generators and complicating dependencies.

hlomzik replied:
Unfortunately the whole styles approach here is too complex anyway. For example, styles are changed by the method and colors they rely on are not in `views` for some reason. And colors are multiple values. But I'll add this as an idea to follow up ticket. And I added a todo already about colors and views.

hlomzik committed:
Fix style linting

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14361897145)
> Successfully merged:  create mode 100644 web/libs/editor/tests/integration/e2e/control_tags/from-prediction.cy.ts

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1973/adjustable-spans'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14361897145

hlomzik committed:
Merge branch 'fb-leap-1973/adjustable-spans' of github.com:HumanSignal/label-studio into fb-leap-1973/adjustable-spans

MihajloHoma commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14374942516)
> Successfully merged:  create mode 100644 web/libs/storybook/addons/theme-toggle/theme-toggle.tsx

MihajloHoma committed:
Merge branch 'develop' into 'fb-leap-1973/adjustable-spans'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14374942516

MihajloHoma commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14374987905)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14374987905

hlomzik committed:
Fix imports from apps into libs to fix tests

We should not use anything from `apps` in `libs`

hlomzik committed:
Another attempt to fix tests via moduleNameMapper

hlomzik committed:
Fix linting in jest config %)

hlomzik committed:
Return imports back, but leave the todo comment

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14381797285)
> Successfully merged: Already up to date.

hlomzik committed:
Fix linting again

hlomzik commented:
@Gondragos thank you, I added link to your comments to follow-up ticket (LEAP-2000)

hlomzik commented:
/help
> Command | Description
> --- | ---
> /frontend [\<args\> ...] | Actions with frontend. Type `/frontend help` for an additional help.

hlomzik committed:
Let's try different kind of imports

hlomzik commented:
> Could you add a screenshot of what this looks like?

Added a video to PR description, thank you for call out!
